### ISSUE TITLE
Uncaught TypeError: Cannot read property 'map' of undefined

### ISSUE THREAD

**Issue by vincentbel** [t9md/atom-project-folder#9]
### Steps to Reproduce

1. Open atom empty(no project opened)
2. Open project folder with shortcut(`cmd + p` for me)
3. Press `space` key to open a new project

**Atom**: 1.16.0 x64
**Electron**: 1.3.13
**OS**: Mac OS X 10.12.4
**Thrown From**: [project-folder](https://github.com/t9md/atom-project-folder) package 1.2.1

### Stack Trace

Uncaught TypeError: Cannot read property 'map' of undefined

```
At /Users/VincentBel/dev/dotfiles/.atom/packages/project-folder/lib/view.coffee:93

TypeError: Cannot read property 'map' of undefined
    at View.module.exports.View.getItems (/packages/project-folder/lib/view.coffee:93:16)
    at View.module.exports.View.confirmAndContinue (/packages/project-folder/lib/view.coffee:112:15)
    at /packages/project-folder/lib/view.coffee:30:50)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/app.asar/src/command-registry.js:259:29)
    at /app.asar/src/command-registry.js:3:59
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:599:16)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:390:22)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (/app.asar/src/window-event-handler.js:106:36)
    at /app.asar/src/window-event-handler.js:3:59)
```

### Commands

```
     -0:16.8.0 core:close (input.hidden-input)
     -0:15.1.0 project-folder:add (atom-pane.pane.active)
     -0:10.9.0 project-folder:confirm-and-continue (input.hidden-input)
     -0:10.8.0 tree-view:show (atom-workspace.workspace.scrollbars-visible-when-scrolling.theme-one-dark-vivid-syntax.theme-one-dark-ui)
     -0:10.8.0 tree-view:toggle-focus (atom-workspace.workspace.scrollbars-visible-when-scrolling.theme-one-dark-vivid-syntax.theme-one-dark-ui)
```

### Non-Core Packages

```
activate-power-mode 1.4.0 
atom-beautify 0.29.23 
atom-material-syntax 1.0.2 
atom-material-ui 1.3.10 
atom-ternjs 0.18.2 
atom-typescript 11.0.3 
autocomplete-emojis 2.5.0 
autocomplete-json 5.4.0 
autocomplete-modules 1.6.9 
base16-ocean-dark-spacegray 0.13.0 
busy-signal 1.4.1 
chester-atom-syntax 0.1.1 
clip-history 0.4.0 
clipboard-plus 0.5.1 
cursor-history 0.10.0 
dash 1.7.1 
docblockr 0.9.4 
file-icons 2.1.4 
hyperclick 0.0.40 
intentions 1.1.2 
ionide-fsharp 1.9.3 
js-hyperclick 1.11.0 
language-babel 2.57.7 
language-docker 1.1.7 
language-elm 1.5.0 
language-haskell 1.12.1 
language-rust 0.4.10 
language-viml 1.1.3 
linter 2.1.4 
linter-elm-make 0.23.2 
linter-eslint 8.1.7 
linter-rust 0.8.0 
linter-spell 0.13.1 
linter-spell-javascript 0.7.2 
linter-stylelint 3.10.0 
linter-ui-default 1.2.3 
local-history 4.0.1 
markdown-preview-plus 2.4.9 
narrow 0.40.1 
nucleus-dark-ui 0.12.1 
one-dark-vivid-syntax 1.7.1 
open-recent 5.0.0 
pigments 0.39.1 
prettier-atom 0.29.0 
project-folder 1.2.1 
racer 0.20.0 
react-function-to-class 0.2.1 
relative-numbers 0.7.1 
spacegray-atom-dark-syntax 1.3.2 
spacegray-dark-neue-syntax 1.4.2 
spacegray-dark-ui 0.13.0 
stylefmt 4.1.1 
todo-show 1.11.0 
tomorrow-night-eighties-syntax 1.0.2 
vim-mode-plus 0.91.0 
```

**Comment by t9md** [comment]
OK, thanks for reporting with the procedure of how-to-reproduce.
Will look this.

**Comment by t9md** [comment]
From my investigation this is not happens in Atom-stable v1.16.0.
And happens in Atom-beta v1.17.x. Because of following steps.

1. When removed project folder from project list, `on-did-change-path` event fired.
2. Then tree-view's `createOrDestroyTreeViewIfNeeded` called.
3. In `createOrDestroyTreeViewIfNeeded` function, TreeView is destroyed when project-list was empty( `atom.project.getPaths().length === 0` ).
4. And remove dock where treeView was lived.( I believe this is new behavior from v1.17.0 ), in previous version treeView was in panel item. and panel was not destroyed.
5. When doc was removed. It **statically re-focus to workspaceCenter's active pane** ( ignore originally focused element ).

https://github.com/atom/atom/pull/14143/commits/73f04fdaed30b774d1f5134a1df05eba30202aab

6. At this point, focus was taken away from select-list's editorElement to workspaceCenter's active pane.
7. So select-list's editor was `blur`-ed and `cancel` was automatically called.

Summarize what is bad.

- When package author get emptied project-list via `atom.project.removePath(dir)`
- It destroy dock of tree-view and original focus is taken away. Since when dock was removed, it always focus workspaceCenter's active pane.
- I want atom code respect original focus.

The cause is clear but fix is not easy, I can add guard to not throw exception, but I cannot prevent cancellation of select-list.
Since as package author-side, cannot prevent `blur` event happened on select-list in normal way.

hmmm.

**Comment by vincentbel** [comment]
@t9md Thanks for the quick investigation.

I'm not familiar with atom development. But the issue do happened in `Atom: 1.16.0 x64` in my machine.(May affected by other packages)

Here is a screenshot:

![project-folder-open-bug](https://cloud.githubusercontent.com/assets/6076919/25646056/e95e0a72-2fe6-11e7-90e0-85c226151d62.gif)

**Comment by t9md** [comment]
Ok, what I explained case is only happen in `remove` case.

But I understand it's also happen in `add` case and also happen in v1.16.0 from your screenshot.
For `space` keymap it add/remove projectPath and keep focus on mini-editor.
This issue happens mini-editor is `canceled` which is caused by change activeElement.
So what I want to know clearly is what/who cause activeElement change when you hit `space`.
Maybe trigger is tree-view creation since it's only happen when you start project-folder with empty projectPaths.

**Comment by vincentbel** [comment]
Got it, thanks for clarification.

**Comment by t9md** [comment]
If you can, could you help me to understand **who-cause-cancel**?
I added small `console.trace()` debug print on `canceled` function.

Want to know `console.trace()` output after following steps below.

## Setup and launch atom

```Shell
cd ~/github/
git clone https://github.com/t9md/atom-project-folder
cd atom-project-folder
git checkout debug-cancelation-from-outside
git pull
apm install

# Create symlink to this directory to make this repo effective when atom was dev-mode.
apm link -d .

# lanuch atom in dev-mode
atom -d
```

## Reproduce issue and see who cause cancel

1. Open chorme-dev-toos by `opt-cmd-i`
2. Reproduce issue.
3. See output of `contole.trace()` on chrome-dev-tool.
4. It would show **who-cause-cacnel**.

## Cleanup

```Shell
# remove symlink to make Atom use normally installed project-folder.
apm unlink -d .
```

**Comment by vincentbel** [comment]
@t9md of course.

![image](https://cloud.githubusercontent.com/assets/6076919/25647718/15f88ab6-2ff7-11e7-940a-5b26fc3613c3.png)

If you need to view the sources of some of the stack trace, just asks.

> the `TextEditorElement.inputNodeBlurred	@	text-editor-element.coffee:155` may related:
>
> ![image](https://cloud.githubusercontent.com/assets/6076919/25647873/83a60baa-2ff8-11e7-9aed-6243207ecc3b.png)

**Comment by t9md** [comment]
Thanks, I understand what caused focus change for your issue is TreeView too.
I believe I should be able to reproduce by my side. 

Do you have any specific setting for tree-view?

When project get available from empty to one(by hitting `space`), tree-view get automatically focused. that cause select-list of atom-project-folder canceled, then fire exception.

Will try to reproduce first by my side.
Thanks for your help.

**Comment by vincentbel** [comment]
@t9md Here is my `tree-view` config:

```cson
"tree-view":
  autoReveal: true
  focusOnReveal: false
  hideIgnoredNames: true
  squashDirectoryNames: true
```

**Comment by vincentbel** [comment]
And the config ui:

![image](https://cloud.githubusercontent.com/assets/6076919/25648062/524fce54-2ffa-11e7-9f67-9aa4403b5dd8.png)

**Comment by t9md** [comment]
thanks, I tried with your config with atom1.16.0 with empty project-list.
But strangely can't reproduce.

From your GIF, when you added very first project-folder by `space`, tree-view left-panel is automatically opened and focus was taken away to tree-view.

But my environment tree-view panel is automatically opened but focus is still on select-list so  can continue add/remove without seeing exception.

Will try to read TreeView code with the info you pasted.

**Comment by t9md** [comment]
Found the diff, and solution. you use file-icons package, if you disable file-icons package.

Your issue will disappear.

Want you confirm before I start to think how to solve this issue.

**Comment by vincentbel** [comment]
@t9md yeah! The issue disappeared if I disable `file-icons` package.

**Comment by t9md** [comment]
Forgot to say i could reproduce issue by installing file-icons. Your screen capture of console.trace give me a hint.

**Comment by t9md** [comment]
Ok will look into and investigate next action.
I know this package use select-list very ubnormally. Most package author expects that slect-list is OK to close after some item confirmed. 
But actually it's not OK for this package.

Let them understand background is a bit tedious.
Maybe i can temporarily disable `bulr` event and take back focus when taken away.
But i think this is too hacky, i want avoid if possible.

**Comment by t9md** [comment]
Opened issue on `file-icons` project.
And as I explained in that issue, following setting can be used as workaround.
This is hidden setting, you need to set via console.log or manually edit config.cson.

```
  "file-icons":
    revealTreeView: false
```

**Comment by vincentbel** [comment]
@t9md Thanks!

With the config added to `config.cson`, the issue disappeared, but it won't show `tree-view` panel automatically.

**Comment by t9md** [comment]
yes, that is strange diff I reported to files-icons project. without installing file-icons, tree-views automatically appear, so I think file-icons have some reason for that diff.

**Comment by t9md** [comment]
Try just released v1.3.0.

- It fixed your issue with file-icons(no workaround command is necessary).
- And also fixed issue I originally investigated and reported to atom-core(now no need upstream fix).

It's a bit hacky but I think this hack still valuable to avoid similar issue in future.
